crates/holochain/src/core/workflow/app_validation_workflow/tests.rs

```diff
@@ -1,41 +1,156 @@
 use holochain_state::test_utils::test_db_dir;
+use holochain_state::validation_db::ValidationStage;
+use holochain_types::dht_op::{DhtOp, DhtOpHashed};
 use holochain_types::inline_zome::InlineZomeSet;
 use holochain_types::prelude::*;
-use holochain_wasm_test_utils::TestWasm;
-
-use holochain_sqlite::error::DatabaseResult;
-use holochain_wasm_test_utils::TestWasmPair;
-use holochain_wasm_test_utils::TestZomes;
+use holochain_wasm_test_utils::{TestWasm, TestWasmPair, TestZomes};
+use holochain_zome_types::action::Dna;
+use holochain_zome_types::fixt::SignatureFixturator;
+use holochain_zome_types::timestamp::Timestamp;
+use holochain_zome_types::Action;
 use matches::assert_matches;
-use rusqlite::named_params;
-use rusqlite::Transaction;
-use std::convert::TryFrom;
-use std::convert::TryInto;
+use rusqlite::{named_params, Transaction};
+use std::convert::{TryFrom, TryInto};
+use std::sync::Arc;
 use std::time::Duration;
 
+#[cfg(test)]
+#[tokio::test(flavor = "multi_thread")]
+async fn main_loop_app_validation_workflow() {
+    use holochain_sqlite::error::DatabaseError;
+    use rusqlite::params;
+
+    holochain_trace::test_run().ok();
+
+    let (dna_file, _, _) = SweetDnaFile::unique_from_test_wasms(vec![TestWasm::Validate]).await;
+    let dna_hash = dna_file.dna_hash().clone();
+
+    let mut conductor = SweetConductor::from_config_rendezvous(
+        SweetConductorConfig::rendezvous(true),
+        SweetLocalRendezvous::new().await,
+    )
+    .await;
+    let app = conductor.setup_app("", &[dna_file.clone()]).await.unwrap();
+    let cell_id = app.cells()[0].cell_id().clone();
+
+    let app_validation_workspace = Arc::new(AppValidationWorkspace::new(
+        conductor.get_authored_db(&dna_hash).unwrap().into(),
+        conductor.get_dht_db(&dna_hash).unwrap(),
+        conductor.get_dht_db_cache(&dna_hash).unwrap(),
+        conductor.get_cache_db(&cell_id).await.unwrap(),
+        conductor.keystore(),
+        Arc::new(dna_file.dna_def().clone()),
+    ));
+    // check there are no ops to app validate
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap();
+    assert_eq!(ops_to_validate.len(), 0);
+
+    // create ops to validate
+    let action = Action::Dna(Dna {
+        author: fixt!(AgentPubKey),
+        timestamp: Timestamp::now(),
+        hash: dna_hash.clone(),
+    });
+    let dht_op = DhtOp::RegisterAgentActivity(fixt!(Signature), action.clone());
+    let dht_op_hash = DhtOpHash::with_data_sync(&dht_op);
+    let dht_op_hashed = DhtOpHashed::with_pre_hashed(dht_op, dht_op_hash.clone());
+
+    let dht_op_2 = DhtOp::StoreEntry(fixt!(Signature), fixt!(NewEntryAction), fixt!(Entry));
+    let dht_op_hash_2 = DhtOpHash::with_data_sync(&dht_op_2);
+    let dht_op_hashed_2 = DhtOpHashed::with_pre_hashed(dht_op_2, dht_op_hash_2.clone());
+
+    app_validation_workspace
```

Comment:
Probably worth extracting to a function, this is something other tests will want too? Guess that can be done when it's needed though :)

Reply:
Yes, I plan to go in many small steps about this refactor. I don't even know if I'll want to keep this test once all the other ones are in place =)